### PR TITLE
Fix for fulfilling an order when product quantity equals allocated quantity

### PR DESCRIPTION
**Pull request by GrzegorzDerdak**
Fix for fulfilling an order when product quantity equals allocated quantity

# Impact

* [ ] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [x] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

# Related frontend Pull Request

https://github.com/mirumee/saleor-dashboard/pull/788

### CODE REVIEW

File: saleor/graphql/order/types.py [hunk 1]
```diff
@@ -299,6 +302,10 @@ def resolve_translated_product_name(root: models.OrderLine, _info):
     def resolve_translated_variant_name(root: models.OrderLine, _info):
         return root.translated_variant_name
 
+    @staticmethod
+    def resolve_allocations(root: models.OrderLine, _info):
+        return root.allocations.all()
```
Comment (fowczarek) [inline saleor/graphql/order/types.py:307]: We should add dataloader here. 
:police_officer:

File: saleor/graphql/order/types.py [hunk 2]
```diff
@@ -299,6 +302,10 @@ def resolve_translated_product_name(root: models.OrderLine, _info):
     def resolve_translated_variant_name(root: models.OrderLine, _info):
         return root.translated_variant_name
 
+    @staticmethod
+    def resolve_allocations(root: models.OrderLine, _info):
```
Comment (fowczarek) [inline saleor/graphql/order/types.py:306]: Permissions?
This field should be visible to staff users who create fulfillment.

File: saleor/graphql/order/types.py [hunk 3]
```diff
@@ -255,6 +255,9 @@ class OrderLine(CountableDjangoObjectType):
     translated_variant_name = graphene.String(
         required=True, description="Variant name in the customer's language"
     )
+    allocations = graphene.List(
+        Allocation, description="List of allocations across warehouses.",
```
Comment (fowczarek) [inline saleor/graphql/order/types.py:259]:
```suggestion
        graphene.NonNull(Allocation), description="List of allocations across warehouses.",
```
We shouldn't allot to return the list with nulls.

File: saleor/graphql/order/types.py (hunk 2 above)
Comment (maarcingebala) [reply saleor/graphql/order/types.py:306]: I would say we should require either `MANGE_ORDERS` or `MANAGE_PRODUCTS`:
```
@one_of_permissions_required([ProductPermissions.MANAGE_PRODUCTS, OrderPermissions.MANAGE_ORDERS])
```

File: saleor/graphql/order/dataloaders.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from collections import defaultdict
+
+from ...warehouse.models import Allocation
+from ..core.dataloaders import DataLoader
+
+
+class AllocationsByOrderLineIdLoader(DataLoader):
+    context_key = "allocations_by_orderline_id"
+
+    def batch_load(self, keys):
```
Comment (GrzegorzDerdak) [inline saleor/graphql/order/dataloaders.py:10]: @fowczarek @maarcingebala please take a look at this DataLoader

File: saleor/graphql/order/dataloaders.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        allocations = Allocation.objects.filter(order_line__pk__in=keys).select_related(
+            "order_line", "stock"
+        )
+        order_lines_to_allocations = defaultdict(list)
+
+        for allocation in allocations:
+            for order_line_id in keys:
+                if allocation.order_line_id == order_line_id:
+                    order_lines_to_allocations[order_line_id].append(allocation)
```
Comment (IKarbowiak) [inline saleor/graphql/order/dataloaders.py:19]: You don't have to iterate through `keys` here. You can create `order_lines_to_allocations` map only with the use of allocations. In the `return` sentence you are choosing values that match the keys. 
```suggestion
        for allocation in allocations:
            order_lines_to_allocations[allocation.order_line_id].append(allocation)
```

File: saleor/graphql/order/dataloaders.py (hunk 4 above)
Comment (maarcingebala) [reply saleor/graphql/order/dataloaders.py:10]: Looks good 👍

**Review by IKarbowiak — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

File: saleor/graphql/order/dataloaders.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+from ..core.dataloaders import DataLoader
+
+
+class AllocationsByOrderLineIdLoader(DataLoader):
+    context_key = "allocations_by_orderline_id"
+
+    def batch_load(self, keys):
+        allocations = Allocation.objects.filter(order_line__pk__in=keys).select_related(
+            "order_line", "stock"
+        )
```
Comment (fowczarek) [inline saleor/graphql/order/dataloaders.py:13]: Why do you use `select_related` here?

File: saleor/graphql/order/dataloaders.py (hunk 5 above)
Comment (fowczarek) [inline saleor/graphql/order/dataloaders.py]:
```suggestion
        for allocation in allocations:
            order_lines_to_allocations[allocation.order_line_id].append(allocation)
```

Comment (IKarbowiak) [reply saleor/graphql/order/dataloaders.py]: @fowczarek It's already done in this way. Refresh changes.

File: saleor/graphql/order/types.py [hunk 7]
```diff
@@ -303,8 +305,11 @@ def resolve_translated_variant_name(root: models.OrderLine, _info):
         return root.translated_variant_name
 
     @staticmethod
+    @one_of_permissions_required(
+        [ProductPermissions.MANAGE_PRODUCTS, OrderPermissions.MANAGE_ORDERS]
+    )
     def resolve_allocations(root: models.OrderLine, _info):
-        return root.allocations.all()
+        return AllocationsByOrderLineIdLoader(_info.context).load(root.id)
```
[the remark this answers is not in the document]
Comment (GrzegorzDerdak) [reply saleor/graphql/order/types.py:312]: I know, but I'd rather stick to the convention from above

[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/order/types.py:312]: `_` prefix means that the arg is not used but it's required by a signature. If it is used like in your case, it should be changed to `info`.

**Pull request merged**